[FLINK-28471]Translate hive_read_write.md into Chinese (#20227)
apache/flink: docs/content.zh/docs/connectors/table/hive/hive_read_write.md

under the License.
-->

# Hive 读 & 写

> 通过使用 `HiveCatalog`，Apache Flink 可以对 Apache Hive 表做统一的批和流处理。这意味着 Flink 可以成为 Hive 批处理引擎的一个性能更好的选择，或者连续读写 Hive 表中的数据以支持为实时数据仓库应用。

Reviewer: ```suggestion
通过使用 `HiveCatalog`，Apache Flink 可以对 Apache Hive 表做统一的批和流处理。这意味着 Flink 可以成为 Hive 批处理引擎的一个性能更好的选择，或者连续读写 Hive 表中的数据以支持实时数据仓库应用。
```

## 读

Flink 支持以批和流两种模式从 Hive 表中读取数据。批读的时候，Flink 会基于执行查询时表的状态进行查询。流读时将持续监控表，并在表中新数据可用时进行增量获取，默认情况下，Flink 将以批模式读取数据。

流读支持消费分区表和非分区表。对于分区表，Flink 会监控新分区的生成，并且在数据可用的情况下增量获取数据。对于非分区表，Flink 将监控文件夹中新文件的生成，并增量地读取新文件。

<table class="table table-bordered">
  <thead>
    <tr>
        <th class="text-left" style="width: 20%">键</th>
        <th class="text-left" style="width: 15%">默认值</th>
        <th class="text-left" style="width: 10%">类型</th>
        <th class="text-left" style="width: 55%">描述</th>
    </tr>
  </thead>
  <tbody>
    <tr>
        <td><h5>streaming-source.enable</h5></td>
        <td style="word-wrap: break-word;">false</td>
        <td>Boolean</td>
        <td>是否启动流读。注意：请确保每个分区/文件都应该原子地写入，否则读取不到完整的数据。</td>
    </tr>
    <tr>
        <td><h5>streaming-source.partition.include</h5></td>
        <td style="word-wrap: break-word;">all</td>
        <td>String</td>

>         <td>选择读取的分区，可选项为 `all` 和 `latest`，`all` 读取所有分区；`latest` 读取按照 'streaming-source.partition.order' 排序后的最新分区，`latest` 仅在 streaming 模式的 hive source table 作为时态表时有效。默认的选项是`all`。在开启 'streaming-source.enable' 并设置 'streaming-source.partition.include' 为 'latest' 时，Flink 支持 temporal join 最新的hive分区，同时，用户可以通过配置分区相关的选项来分配分区比较顺序和数据更新时间间隔。 

Reviewer: ```suggestion
        <td>选择读取的分区，可选项为 `all` 和 `latest`，`all` 读取所有分区；`latest` 读取按照 'streaming-source.partition.order' 排序后的最新分区，`latest` 仅在流模式的 Hive 源表作为时态表时有效。默认的选项是`all`。在开启 'streaming-source.enable' 并设置 'streaming-source.partition.include' 为 'latest' 时，Flink 支持 temporal join 最新的 Hive 分区，同时，用户可以通过配置分区相关的选项来配置分区比较顺序和数据更新时间间隔。 
```

        </td>
    </tr>     
    <tr>
        <td><h5>streaming-source.monitor-interval</h5></td>
        <td style="word-wrap: break-word;">None</td>
        <td>Duration</td>
        <td>连续监控分区/文件的时间间隔。

>             注意: 默认情况下，流式读 Hive 的间隔为 '1 min'，但流读 Hive 的 temporal join 的默认时间间隔是 '60 min'，这是因为当前流读 Hive 的 temporal join 实现上有一个框架限制，即每个 TM 都要访问 Hive metaStore，这可能会对 metaStore 产生压力，这个问题将在未来得到改善。</td>

Reviewer: ```suggestion
            注意: 默认情况下，流式读 Hive 的间隔为 '1 min'，但流读 Hive 的 temporal join 的默认时间间隔是 '60 min'，这是因为当前流读 Hive 的 temporal join 实现上有一个框架限制，即每个 TM 都要访问 Hive metastore，这可能会对 metastore 产生压力，这个问题将在未来得到改善。</td>
```

>             注意: 默认情况下，流式读 Hive 的间隔为 '1 min'，但流读 Hive 的 temporal join 的默认时间间隔是 '60 min'，这是因为当前流读 Hive 的 temporal join 实现上有一个框架限制，即每个TM都要访问 Hive metaStore，这可能会对metaStore产生压力，这个问题将在未来得到改善。</td>

Reviewer: ```suggestion
            注意: 默认情况下，流式读 Hive 的间隔为 '1 min'，但流读 Hive 的 temporal join 的默认时间间隔是 '60 min'，这是因为当前流读 Hive 的 temporal join 实现上有一个框架限制，即每个TM都要访问 Hive metastore，这可能会对metastore产生压力，这个问题将在未来得到改善。</td>
```

    </tr>
    <tr>
        <td><h5>streaming-source.partition-order</h5></td>
        <td style="word-wrap: break-word;">partition-name</td>
        <td>String</td>

>         <td>streaming source 分区排序，支持 create-time， partition-time 和 partition-name。 create-time 比较分区/文件创建时间， 这不是 Hive metaStore 中创建分区的时间，而是文件夹/文件在文件系统的修改时间，如果分区文件夹以某种方式更新，比如添加在文件夹里新增了一个文件，它会影响到数据的使用。partition-time 从分区名称中抽取时间进行比较。partition-name 会比较分区名称的字典顺序。对于非分区的表，总是会比较 'create-time'。对于分区表默认值是 'partition-name'。该选项与已经弃用的 'streaming-source.consume-order' 的选项相同</td>

Reviewer: ```suggestion
        <td>streaming source 分区排序，支持 create-time， partition-time 和 partition-name。 create-time 比较分区/文件创建时间， 这不是 Hive metastore 中创建分区的时间，而是文件夹/文件在文件系统的修改时间，如果分区文件夹以某种方式更新，比如添加在文件夹里新增了一个文件，它会影响到数据的使用。partition-time 从分区名称中抽取时间进行比较。partition-name 会比较分区名称的字典顺序。对于非分区的表，总是会比较 'create-time'。对于分区表默认值是 'partition-name'。该选项与已经弃用的 'streaming-source.consume-order' 的选项相同</td>
```

    </tr>
    <tr>
        <td><h5>streaming-source.consume-start-offset</h5></td>
        <td style="word-wrap: break-word;">None</td>
        <td>String</td>

>         <td>流模式起始消费偏移量。如何解析和比较偏移取决于你指定的顺序。对于 create-time 和 partition-time，会比较字符串类型的 timestamp (yyyy-[m]m-[d]d [hh:mm:ss])。对于 partition-time，将使用分区时间提取器从分区提取的时间。

Reviewer: ```suggestion
        <td>流模式起始消费偏移量。如何解析和比较偏移量取决于你指定的顺序。对于 create-time 和 partition-time，会比较时间戳 (yyyy-[m]m-[d]d [hh:mm:ss])。对于 partition-time，将使用分区时间提取器从分区名字中提取的时间。
```

```

**注意**

- 监控策略是扫描当前位置路径中的所有目录/文件，分区太多可能导致性能下降。
- 流读非分区表时要求每个文件应原子地写入目标目录。
- 流读分区表要求每个分区应该被原子地添加进 Hive metastore 中。如果不是的话，只有添加到现有分区的新数据会被消费。 

> - 流读 Hive 表不支持Flink DDL 的 watermark 语法。这些表不能被用于窗口算子。

Reviewer: ```suggestion
- 流读 Hive 表不支持 Flink DDL 的 watermark 语法。这些表不能被用于窗口算子。
```

### 读取 Hive Views

Flink 能够读取 Hive 中已经定义的视图。但是也有一些限制：

1) Hive catalog 必须设置成当前的 catalog 才能查询视图。在 Table API 中使用 `tableEnv.useCatalog(...)`，或者在 SQL 客户端使用 `USE CATALOG ...` 来改变当前 catalog。

> 2) Hive 和 Flink SQL 的语法不同, 比如不同的关键字和字面值。确保查询视图与 Flink 语法兼容。

Reviewer: ```suggestion
2) Hive 和 Flink SQL 的语法不同, 比如不同的关键字和字面值。请确保对视图的查询语法与 Flink 语法兼容。
```

### 读取时的向量化优化

> 当满足以下条件时，Flink会自动对Hive table进行向量化读取:

Reviewer: ```suggestion
当满足以下条件时，Flink 会自动对 Hive 表进行向量化读取:
```

- 格式：ORC 或者 Parquet。
- 没有复杂类型的列，比如 Hive 列类型：List，Map，Struct，Union。

该特性默认开启， 可以使用以下配置禁用它。

```bash
table.exec.hive.fallback-mapred-reader=true
```

### Source 并发推断

默认情况下，Flink 会基于文件的数量，以及每个文件中块的数量推断出读取 Hive 的最佳并行度。

> Flink 允许你灵活的配置并发推断策略。你可以在 `TableConfig` 中配置以下参数(注意这些参数会影响当前作业的所有 source )：

Reviewer: ```suggestion
Flink 允许你灵活的配置并发推断策略。你可以在 `TableConfig` 中配置以下参数(注意这些参数会影响当前作业的所有 source)：
```

<table class="table table-bordered">
  <thead>
    <tr>
        <th class="text-left" style="width: 20%">键</th>
        <th class="text-left" style="width: 15%">默认值</th>
        <th class="text-left" style="width: 10%">类型</th>
        <th class="text-left" style="width: 55%">描述</th>
    </tr>
  </thead>
  <tbody>
    <tr>
        <td><h5>table.exec.hive.infer-source-parallelism</h5></td>
        <td style="word-wrap: break-word;">true</td>
        <td>Boolean</td>
        <td>如果是 true，会根据 split 的数量推断 source 的并发度。如果是 false，source 的并发度由配置决定。</td>
    </tr>
    <tr>
        <td><h5>table.exec.hive.infer-source-parallelism.max</h5></td>
        <td style="word-wrap: break-word;">1000</td>
        <td>Integer</td>
        <td>设置 source operator 推断的最大并发度。</td>
    </tr>
  </tbody>
</table>

### 加载分区切片

Flink 使用多个线程并发将 Hive 分区切分成多个 split 进行读取。你可以使用 `table.exec.hive.load-partition-splits.thread-num` 去配置线程数。默认值是3，你配置的值应该大于0。

### 读取带有子目录的分区

> 在某些情况下，你可以创建一个引用其他表的外部表，但是该表的分区列是另一张表分区字段的子集。

Reviewer: ```suggestion
在某些情况下，你或许会创建一个引用其他表的外部表，但是该表的分区列是另一张表分区字段的子集。
```

比如，你创建了一个分区表 `fact_tz` ，分区字段是 `day` 和 `hour` ：

```sql
CREATE TABLE fact_tz(x int) PARTITIONED BY (day STRING, hour STRING);
```

> 当你使用 `fact_tz` 表创建了一个外部表 `fact_daily` ，并使用了一个粗粒度的分区字段 `day`：

Reviewer: ```suggestion
然后你基于 `fact_tz` 表创建了一个外部表 `fact_daily` ，并使用了一个粗粒度的分区字段 `day`：
```

```sql
CREATE EXTERNAL TABLE fact_daily(x int) PARTITIONED BY (ds STRING) LOCATION '/path/to/fact_tz';
```

当读取外部表 `fact_daily` 时，该表的分区目录下存在子目录(`hour=1` 到 `hour=24`)。

默认情况下，可以将带有子目录的分区添加到外部表中。Flink SQL 会递归扫描所有的子目录，并获取所有子目录中数据。

```sql
ALTER TABLE fact_daily ADD PARTITION (ds='2022-07-07') location '/path/to/fact_tz/ds=2022-07-07';
```

> 你可以设置作业属性 `table.exec.hive.read-partition-with-subdirectory.enabled` (默认时 `true` ) 为 `false` 以禁止 Flink 读取子目录。

Reviewer: ```suggestion
你可以设置作业属性 `table.exec.hive.read-partition-with-subdirectory.enabled` (默认为 `true` ) 为 `false` 以禁止 Flink 读取子目录。
```

> 如果你设置成 `false` 并且分区目录下不包含文件，而是由子目录组成，Flink 会抛出 `java.io.IOException: Not a file: /path/to/data/*` 异常。

Reviewer: ```suggestion
如果你设置成 `false` 并且分区目录下包含任何子目录，Flink 会抛出 `java.io.IOException: Not a file: /path/to/data/*` 异常。
```

```

**注意**

- 监控策略是扫描当前位置路径中的所有目录/文件，分区太多可能导致性能下降。
- Streaming 读取非分区表，要求每个文件应原子地写入目标目录。

> - Streaming 读取分区表，要求原子地在 Hive metastore 视图中添加分区。如果没有，添加到现有分区的新数据将被消费。 

Reviewer: ```suggestion
- 流读分区表要求每个分区应该被原子地添加进 Hive metastore 中。如果不是的话，只有添加到现有分区的新数据会被消费。 
```

- Streaming 读取不支持Flink DDL的水位线语法。这些表不能被用于 window operators。

### 读取 Hive Views

Flink 能够读取Hive中已经定义的视图。但是也有一些限制：

1) Hive catalog 必须设置成当前的 catalog 才能查询视图。在Table API中使用 `tableEnv.useCatalog(...)`，或者在SQL客户端使用`USE CATALOG ...`来改变当前catalog。

2) Hive 和Flink SQL 的语法不同, 比如不同的关键字和字面值。确保查询视图与Flink语法兼容。

### 读取时的矢量优化

当满足以下条件时，Flink会自动对Hive table进行向量化读取:

- 格式：ORC 或者 Parquet。
- 没有复杂类型的列，像hive类型：List，Map，Struct，Union。

该特性默认开启， 可以使用以下配置禁用它。

```bash
table.exec.hive.fallback-mapred-reader=true
```

> ### Source Parallelism Inference

Reviewer: ```suggestion
### Source 并发 推断
```

## Temporal Table Join

你可以使用 Hive table 作为 temporal table，然后 stream 就可以使用 temporal join 关联 Hive table。
请参照 [temporal join]({{< ref "docs/dev/table/sql/queries/joins" >}}#temporal-joins) 获取更多关于 temporal join 的信息。

Flink 支持 processing-time temporal join Hive Table，processing-time temporal join 总是关联最新版本的 temporal table。
Flink 支持 temporal join Hive 的分区表和非分区表，对于分区表，Flink 支持自动跟踪 Hive table 的最新分区。

> **注意**: Flink 还没有支持 event-time temporal join Hive table。

Reviewer: ```suggestion
**注意**: Flink 还不支持 event-time temporal join Hive 表。
```

### Temporal Join The Latest Partition

> 对于随时变化的分区表，我们可以把它读成一个无界流，如果每个分区包含完整数据，则分区作为 temporal table 的一个版本。temporal table 的版本保存分区的数据。

Reviewer: ```suggestion
对于随时变化的分区表，我们可以把它看作是一个无界流进行读取，如果每个分区包含完整数据，则分区可以作为时态表 的一个版本，时态表的版本保存分区的数据。
```

 
Flink 支持在使用 processing time temporal join 时自动追踪最新的分区（版本），通过 `streaming-source.partition-order` 定义最新的分区（版本）。

> 用户最常使用的案例就是在 Flink stream 应用作业中使用 Hive table 作为维度表。

Reviewer: ```suggestion
用户最常使用的案例就是在 Flink 流作业中使用 Hive 表作为维度表。
```

> **注意:** 该特性仅支持 Flink `STREAMING` 模式。

Reviewer: ```suggestion
**注意:** 该特性仅支持 Flink 流模式。
```

> 下面的案例演示了经典的业务 pipeline，使用 Hive 中的表作为维度表，它们由每天一次的批任务或者 Flink 任务来更新。使用 Hive 中每天批处理或者 Flink 任务更新的表作为维度表，kafka stream 来自实时在线业务数据或者日志，需要关联维度表丰富流。

Reviewer: ```suggestion
下面的案例演示了经典的业务 pipeline，使用 Hive 中的表作为维度表，它们由每天一次的批任务或者 Flink 任务来更新。
Kafka 数据流来自实时在线业务数据或者日志，该流需要关联维度表以丰富数据流。
```

  'streaming-source.enable' = 'true',
  'streaming-source.partition.include' = 'latest',
  'streaming-source.monitor-interval' = '12 h',

>   'streaming-source.partition-order' = 'partition-name',  -- 选项是默认的，可以忽略

Reviewer: ```suggestion
  'streaming-source.partition-order' = 'partition-name',  -- 有默认值的配置项，可以不填。
```

) WITH (...);

-- streaming sql, kafka temporal join hive维度表. Flink 将在 'streaming-source.monitor-interval' 的间隔内自动加载最新分区的数据。

SELECT * FROM orders_table AS o 
JOIN dimension_table FOR SYSTEM_TIME AS OF o.proctime AS dim
ON o.product_id = dim.product_id;

```

### Temporal Join 最新的表
 
对于 Hive 表，我们可以把它看作是一个无界流进行读取，在这个案例中，当我们查询时只能去追踪最新的版本。

> 最新版本的表保留了Hive 表的所有数据。

Reviewer: ```suggestion
最新版本的表保留了 Hive 表的所有数据。
```

> 当 temporal join 最新的 Hive 表，Hive 表 会缓存到 Slot 内存中，并且 数据流中的每条记录通过 key 去关联表找到对应的匹配项。

Reviewer: ```suggestion
当 temporal join 最新的 Hive 表，Hive 表会缓存到 Slot 内存中，并且数据流中的每条记录通过 key 去关联表找到对应的匹配项。
```

> 使用最新的 Hive table 作为时态表不需要额外的配置。作为可选项，您可以使用以下配置项配置 Hive 表 缓存的 TTL。当缓存失效，Hive table 会重新扫描并加载最新的数据。

Reviewer: ```suggestion
使用最新的 Hive 表作为时态表不需要额外的配置。作为可选项，您可以使用以下配置项配置 Hive 表缓存的 TTL。当缓存失效，Hive 表会重新扫描并加载最新的数据。
```

<table class="table table-bordered">
  <thead>
    <tr>
        <th class="text-left" style="width: 20%">键</th>
        <th class="text-left" style="width: 15%">默认值</th>
        <th class="text-left" style="width: 10%">类型</th>
        <th class="text-left" style="width: 55%">描述</th>
    </tr>
  </thead>
  <tbody>
    <tr>
        <td><h5>lookup.join.cache.ttl</h5></td>
        <td style="word-wrap: break-word;">60 min</td>
        <td>Duration</td>

>         <td>在 lookup join 时构建表缓存的 TTL (例如 10min)。默认的 TTL 是60分钟。注意: 该选项仅在 lookup 表为有界的 Hive 表时有效，如果你使用流式的 Hive 表 作为时态表，请使用 'streaming-source.monitor-interval' 去配置数据更新的间隔。

Reviewer: ```suggestion
        <td>在 lookup join 时构建表缓存的 TTL (例如 10min)。默认的 TTL 是60分钟。注意: 该选项仅在 lookup 表为有界的 Hive 表时有效，如果你使用流式的 Hive 表作为时态表，请使用 'streaming-source.monitor-interval' 去配置数据更新的间隔。
```

### Temporal Join The Latest Table
 
对于 Hive table，我们可以把它读成一个无界流，在这个案例中，当我们查询时只能去追踪最新的版本。
最新版本的表保留了Hive table 的所有数据。

> 当执行 temporal join 最新的 Hive table，Hive table 会缓存到 Slot 内存中，并且 stream 中的每条记录通过 key 去关联表找到对应的匹配项。

Reviewer: ```suggestion
当 temporal join 最新的 Hive 表，Hive 表 会缓存到 Slot 内存中，并且 数据流中的每条记录通过 key 去关联表找到对应的匹配项。
```

> 使用最新的 Hive table 作为 temporal table 不需要额外的配置。可选项，您可以使用以下属性配置 Hive table 缓存的 TTL。当缓存失效，Hive table 会重新扫描并加载最新的数据。

Reviewer: ```suggestion
使用最新的 Hive table 作为 temporal table 不需要额外的配置。作为可选项，您可以使用以下配置项配置 Hive 表 缓存的 TTL。当缓存失效，Hive table 会重新扫描并加载最新的数据。
```

<table class="table table-bordered">
  <thead>
    <tr>
        <th class="text-left" style="width: 20%">Key</th>
        <th class="text-left" style="width: 15%">默认值</th>
        <th class="text-left" style="width: 10%">类型</th>
        <th class="text-left" style="width: 55%">描述</th>
    </tr>
  </thead>
  <tbody>
    <tr>
        <td><h5>lookup.join.cache.ttl</h5></td>
        <td style="word-wrap: break-word;">60 min</td>
        <td>Duration</td>

>         <td>在 lookup join 时构建表缓存的 TTL (例如 10min)。默认的 TTL 是60分钟。注意: 该选项仅在查找有边界的 hive table source 时有效，如果你使用 streaming hive source 作为 temporal table，请使用 'streaming-source.monitor-interval' 去配置数据更新的间隔。

Reviewer: ```suggestion
        <td>在 lookup join 时构建表缓存的 TTL (例如 10min)。默认的 TTL 是60分钟。注意: 该选项仅在 lookup 表为有界的 Hive 表时有效，如果你使用流式的 Hive 表 作为 temporal table，请使用 'streaming-source.monitor-interval' 去配置数据更新的间隔。
```

  update_user STRING,
  ...
) TBLPROPERTIES (

>   'streaming-source.enable' = 'false',           -- 选项是默认的，可以忽略。

Reviewer: ```suggestion
  'streaming-source.enable' = 'false',           -- 有默认值的配置项，可以不填。
```

>   'streaming-source.partition.include' = 'all',  -- 选项是默认的，可以忽略。

Reviewer: ```suggestion
  'streaming-source.partition.include' = 'all',  -- 有默认值的配置项，可以不填。
```

) WITH (...);

-- streaming sql, kafka join hive维度表. 当缓存失效时 Flink 会加载维度表的所有数据。

SELECT * FROM orders_table AS o 
JOIN dimension_table FOR SYSTEM_TIME AS OF o.proctime AS dim
ON o.product_id = dim.product_id;

```
注意: 

> 1. 每个参加连接的 subtask 需要在他们的缓存中保留 Hive table。请确保 Hive table 可以放到 TM task slot 中。

Reviewer: ```suggestion
1. 每个参与 join 的 subtask 需要在他们的缓存中保留 Hive 表。请确保 Hive table 可以放到 TM task slot 中。
```

) WITH (...);

> -- streaming sql, kafka join hive维度表. 当缓存失效时 Flink 会加载维度表的所有数据。

Reviewer: ```suggestion
-- streaming sql, kafka join Hive 维度表. 当缓存失效时 Flink 会加载维度表的所有数据。
```

SELECT * FROM orders_table AS o 
JOIN dimension_table FOR SYSTEM_TIME AS OF o.proctime AS dim
ON o.product_id = dim.product_id;

```
注意: 

1. 每个参与 join 的 subtask 需要在他们的缓存中保留 Hive 表。请确保 Hive 表可以放到 TM task slot 中。
2. 建议把这两个选项配置成较大的值`streaming-source.monitor-interval`(最新的分区作为时态表) 和 `lookup.join.cache.ttl`(所有的分区作为时态表)。否则，任务会频繁更新和加载表，容易出现性能问题。

> 3. 目前，缓存刷新的时候会重新加载整个Hive 表，使用没有办法区分数据是新数据还是旧数据。

Reviewer: ```suggestion
3. 目前，缓存刷新的时候会重新加载整个Hive 表，所以没有办法区分数据是新数据还是旧数据。
```

```

> 默认情况下，对于 streaming writes，Flink 仅支持重命名 committers，对于S3文件系统不支持 streaming writes 的 exactly-once 语义。

Reviewer: ```suggestion
默认情况下，对于流，Flink 仅支持重命名 committers，对于S3文件系统不支持流写的 exactly-once 语义。
```

通过将以下参数设置为false，可以实现 exactly-once 写入S3。

> 这会调用 Flink 原生的 writer ，但是仅针对 parquet 和 orc 文件类型。

Reviewer: ```suggestion
这会调用 Flink 原生的 writer ，但是仅针对 parquet 和 orc 文件类型有效。
```

> 在`TableConfig`中配置，会影响作业的所有 sink。

Reviewer: ```suggestion
这个配置项可以在 `TableConfig` 中配置，该配置项对作业的所有 sink 都生效。
```

```

> 默认情况下，对于流，Flink 仅支持重命名 committers，对于S3文件系统不支持流写的 exactly-once 语义。

Reviewer: ```suggestion
默认情况下，对于流，Flink 仅支持重命名 committers，对于 S3 文件系统不支持流写的 exactly-once 语义。
```

> 通过将以下参数设置为false，可以实现 exactly-once 写入S3。

Reviewer: ```suggestion
通过将以下参数设置为 false，可以实现 exactly-once 写入S3。
```
